src/hardware/mixer.cpp

```diff
@@ -1383,35 +1419,81 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
+	auto &convert_out = do_resample ? mixer.resample_temp : mixer.resample_out;
 	ConvertSamples<Type, stereo, signeddata, nativeorder>(data, frames, convert_out);
 
-	auto out_frames = check_cast<spx_uint32_t>(convert_out.size()) / 2u;
+	auto out_frames = check_cast<uint32_t>(convert_out.size()) / 2u;
+
+	if (do_resample) {
+		switch (resample_method) {
+		case ResampleMethod::LinearInterpolation: {
+			auto &s = lerp_upsampler;
+
+			auto in_pos  = mixer.resample_temp.begin();
+			auto &out = mixer.resample_out;
+			out.resize(0);
+
+			out_frames = 0;
 
-	if (do_resampler) {
-		auto in_frames = check_cast<spx_uint32_t>(mixer.resample_temp.size()) / 2u;
+			while (in_pos != mixer.resample_temp.end()) {
+				AudioFrame curr_frame = {*in_pos, *(in_pos + 1)};
 
-		out_frames = estimate_max_out_frames(resampler.state, in_frames);
+				const auto out_left = lerp(s.last_frame.left,
+				                           curr_frame.left,
+				                           s.pos);
 
-		mixer.resample_out.resize(out_frames * 2);
+				const auto out_right = lerp(s.last_frame.right,
+				                            curr_frame.right,
+				                            s.pos);
 
-		speex_resampler_process_interleaved_float(resampler.state,
-		                                          mixer.resample_temp.data(),
-		                                          &in_frames,
-		                                          mixer.resample_out.data(),
-		                                          &out_frames);
+				out.emplace_back(out_left);
+				out.emplace_back(out_right);
+				++out_frames;
 
-		// out_frames now contains the actual number of resampled frames,
-		// ensure the number of output frames is within the logical size.
-		assert(out_frames <= mixer.resample_out.size() / 2);
-		mixer.resample_out.resize(out_frames * 2);  // only shrinks
+				s.pos += s.step;
+
+				if (s.pos > 1.0f) {
+					s.pos -= 1.0f;
+					s.last_frame = curr_frame;
+
+					// Move to the next input frame
+					in_pos += 2;
+				}
+			}
+		} break;
+
+		case ResampleMethod::ZeroOrderHoldAndResample: [[fallthrough]];
+		case ResampleMethod::Resample: {
+			auto in_frames = check_cast<uint32_t>(
+			                         mixer.resample_temp.size()) /
+			                 2u;
+
+			out_frames = estimate_max_out_frames(speex_resampler.state,
+			                                     in_frames);
+
+			mixer.resample_out.resize(out_frames * 2);
+
+			speex_resampler_process_interleaved_float(
+			        speex_resampler.state,
+			        mixer.resample_temp.data(),
+			        &in_frames,
+			        mixer.resample_out.data(),
+			        &out_frames);
+
+			// out_frames now contains the actual number of
+			// resampled frames, ensure the number of output frames
+			// is within the logical size.
+			assert(out_frames <= mixer.resample_out.size() / 2);
+			mixer.resample_out.resize(out_frames * 2); // only shrinks
+		} break;
+		}
```

Comment:
Just noticed in the above that we can eliminate the out_frames tallying and initial counts, and instead directly query the `resampled_out` vector directly as we head into the mixing loop.

```diff

diff --git a/src/hardware/mixer.cpp b/src/hardware/mixer.cpp
index 4ec58f97f..ce1591686 100644
--- a/src/hardware/mixer.cpp
+++ b/src/hardware/mixer.cpp
@@ -1422,8 +1422,6 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
        auto &convert_out = do_resample ? mixer.resample_temp : mixer.resample_out;
        ConvertSamples<Type, stereo, signeddata, nativeorder>(data, frames, convert_out);
 
-       auto out_frames = check_cast<uint32_t>(convert_out.size()) / 2u;
-
        if (do_resample) {
                switch (resample_method) {
                case ResampleMethod::LinearInterpolation: {
@@ -1433,8 +1431,6 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
                        auto &out = mixer.resample_out;
                        out.resize(0);
 
-                       out_frames = 0;
-
                        while (in_pos != mixer.resample_temp.end()) {
                                AudioFrame curr_frame = {*in_pos, *(in_pos + 1)};
 
@@ -1448,7 +1444,6 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
 
                                out.emplace_back(out_left);
                                out.emplace_back(out_right);
-                               ++out_frames;
 
                                s.pos += s.step;
 
@@ -1468,8 +1463,8 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
                                                 mixer.resample_temp.size()) /
                                         2u;
 
-                       out_frames = estimate_max_out_frames(speex_resampler.state,
-                                                            in_frames);
+                       auto out_frames = estimate_max_out_frames(
+                               speex_resampler.state, in_frames);
 
                        mixer.resample_out.resize(out_frames * 2);
 
@@ -1493,6 +1488,8 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
 
        // Optionally filter, apply crossfeed, then mix the results to the
        // master output
+       const auto out_frames = static_cast<int>(mixer.resample_out.size()) / 2;
+
        auto pos    = mixer.resample_out.begin();
        auto mixpos = check_cast<work_index_t>(mixer.pos + frames_done);
```

Reply:
(I was thinking we could even eliminate that last `const auto out_frames`, and directly increment the `frames_done` counter inside the mix loop, but it's an atomic, so each increment is a bit more expensive than a normal variable.. so the current bulk-add after the mix loop `frames_done += out_frames;` still makes the most sense.

Reply:
Nice, that simplifies it a bit, and I agree with the second comment as well @kcgen .